### PR TITLE
Core: Minor metadata table code harmonization for readable_metrics

### PR DESCRIPTION
**Pull request by szehon-ho**
Some minor cleanup discussed:  https://github.com/apache/iceberg/pull/7539#discussion_r1192497611 ,

Try to make the readable_metrics code of BaseEntriesTable and BaseFilesTable align, by extracting common class and making code more similar.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseEntriesTable.java [hunk 1]
```diff
@@ -133,16 +133,13 @@ public CloseableIterable<StructLike> rows() {
...
       if (readableMetricsField == null) {
-        CloseableIterable<StructLike> entryAsStruct =
-            CloseableIterable.transform(
-                entries(fileProjection),
-                entry -> (GenericManifestEntry<? extends ContentFile<?>>) entry);
-
         StructProjection structProjection = structProjection(projection);
-        return CloseableIterable.transform(entryAsStruct, structProjection::wrap);
+
+        return CloseableIterable.transform(
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseEntriesTable.java:138]: I was trying to make both branches be more alike:

1. calculate final struct projection
2. calculate 'file' projection if needed for reading manifest
3. Use these to transform the result

File: core/src/main/java/org/apache/iceberg/BaseEntriesTable.java [hunk 2]
```diff
@@ -166,21 +161,35 @@ private Schema requiredFileProjection() {
...
     }
 
+    /**
+     * @param fileStructProjection projection to apply on the 'data_files' struct
+     * @return entries of this read task's manifest
+     */
     private CloseableIterable<? extends ManifestEntry<? extends ContentFile<?>>> entries(
-        Schema newFileProjection) {
-      return ManifestFiles.open(manifest, io, specsById).project(newFileProjection).entries();
+        Schema fileStructProjection) {
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseEntriesTable.java:180]: Make this method a bit more functional (input/output is more clear)

**Comment by szehon-ho** [comment]
@dramaticlly fyi

File: core/src/main/java/org/apache/iceberg/BaseFilesTable.java [hunk 3]
```diff
@@ -192,14 +181,46 @@ private CloseableIterable<? extends ContentFile<?>> files(Schema fileProjection)
...
+     * metrics in human-readable form.
+     *
+     * @file content file metadata
+     * @param readableMetricsField projected "readable_metrics" field
+     * @return struct representing content file, with appended readable_metrics field
+     */
+    private StructLike withReadableMetrics(
+        ContentFile<?> file, Types.NestedField readableMetricsField) {
+      int metricsPosition = projection.columns().indexOf(readableMetricsField);
       int columnCount = projection.columns().size();
```
Comment (dramaticlly) [inline core/src/main/java/org/apache/iceberg/BaseFilesTable.java:195]: maybe we can reuse the `projectionColumnCount` as variable name above ?

File: core/src/main/java/org/apache/iceberg/BaseFilesTable.java [hunk 4]
```diff
@@ -192,14 +181,46 @@ private CloseableIterable<? extends ContentFile<?>> files(Schema fileProjection)
...
+      return new MetricsUtil.StructWithReadableMetrics(
+          (StructLike) file, readableMetrics, columnCount, metricsPosition);
+    }
+
+    /**
+     * Create a projection on content files metadata by removing virtual 'readable_column' and
+     * ensuring that the underlying metrics used to create that column are part of the final
+     * projection.
+     *
+     * @param projectionSchema projection to transform
```
Comment (dramaticlly) [inline core/src/main/java/org/apache/iceberg/BaseFilesTable.java:209]: nit: `projection to transform` read a bit weird as first time I read it I am not clear on why do we need to transform, maybe `requested projection` or `intended projection` which correspond to returning of actual projection ?

**Review by dramaticlly — APPROVED** [review]
thank you @szehon-ho for the refactoring. I think now logic in BaseFiles is much easier to read

File: core/src/main/java/org/apache/iceberg/BaseFilesTable.java (hunk 3 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseFilesTable.java]: Yea, projectionColumnCount is I guess the description of what it is, but the way I think about the method withReadableMetrics is actually  more towards the goal, of making a struct.

So I think it's better to have the variables in the end: struct, structSize, metricsStruct, metricsPosition

(projectionColumnCount being structSize).  Changed the references to match.  What do you think?

File: core/src/main/java/org/apache/iceberg/BaseFilesTable.java (hunk 4 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseFilesTable.java]: I like that, changed to 'requestedProjection'

File: core/src/main/java/org/apache/iceberg/BaseFilesTable.java (hunk 3 above)
Comment (dramaticlly) [reply core/src/main/java/org/apache/iceberg/BaseFilesTable.java]: thanks for the explanation, I think structSize sounds great!

**Comment by szehon-ho** [comment]
Merged, as this is just a continuation/code fixup of #7539.

**Pull request merged**